File: sdk/java/src/test/java/io/megacache/ConformanceTest.java
package io.megacache;

import java.io.OutputStream;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

public final class ConformanceTest {
    @FunctionalInterface
    private interface ClientCall {
        void run(MegaCacheClient client) throws Exception;
    }

    private static Map<String, String> fixtures() throws Exception {
        Map<String, String> values = new LinkedHashMap<>();
        for (String line : Files.readAllLines(Path.of(System.getenv("MEGACACHE_FIXTURES")))) {
            if (line.isEmpty() || line.startsWith("#")) continue;
            String[] parts = line.split("\t", 2); values.put(parts[0], parts[1]);
        }
        return values;
    }
    private static MegaCacheClient client() {
        MegaCacheClient.Options options = new MegaCacheClient.Options();
        options.host = System.getenv("MEGACACHE_CONFORMANCE_HOST");
        options.port = Integer.parseInt(System.getenv("MEGACACHE_CONFORMANCE_PORT"));
        options.invalidationPoll = Duration.ZERO;
        return new MegaCacheClient(options);
    }
    private static void check(boolean condition, String message) {
        if (!condition) throw new AssertionError(message);
    }
    private static void protocolLimitRegression(String response) throws Exception {
        try (ServerSocket server = new ServerSocket(0)) {
            Thread responder = new Thread(() -> {
                try {
                    try (Socket socket = server.accept()) {
                        socket.getInputStream().read();
                        OutputStream output = socket.getOutputStream();
                        output.write(response.getBytes(StandardCharsets.US_ASCII));
                        output.flush();
                    }
                    try (Socket socket = server.accept()) {
                        socket.getInputStream().read();
                        OutputStream output = socket.getOutputStream();
                        output.write("+PONG\r\n".getBytes(StandardCharsets.US_ASCII));
                        output.flush();
                    }
                } catch (Exception error) {
                    throw new RuntimeException(error);
                }
            });
            responder.start();
            MegaCacheClient.Options options = new MegaCacheClient.Options();
            options.host = "127.0.0.1";
            options.port = server.getLocalPort();
            options.timeout = Duration.ofSeconds(1);
            try (MegaCacheClient client = new MegaCacheClient(options)) {
                boolean rejected = false;
                try { client.ping(); }
                catch (MegaCacheClient.ProtocolException expected) { rejected = true; }
                check(rejected, "oversized bulk response was accepted");
                check("PONG".equals(client.ping()), "poisoned connection was reused");
            }
            responder.join();
        }
    }
    private static String bulk(String value) {
        return "$" + value.getBytes(StandardCharsets.UTF_8).length
            + "\r\n" + value + "\r\n";
    }
    private static void malformedResponseRegression(
        String name, String response, ClientCall call
    ) throws Exception {
        try (ServerSocket server = new ServerSocket(0)) {
            Thread responder = respondOnce(server, response, 0);
            try (MegaCacheClient client = deadlineClient(server.getLocalPort())) {
                boolean rejected = false;
                try { call.run(client); }
                catch (MegaCacheClient.ProtocolException expected) {
                    rejected = true;
                }
                check(rejected, name + " did not raise ProtocolException");
            }
            responder.join();
        }
    }
    private static void authExceptionRegressions() throws Exception {
        for (boolean protocol : new boolean[]{false, true}) {
            try (ServerSocket server = new ServerSocket(0)) {
                Thread responder = respondOnce(
                    server,
                    protocol ? ":1\r\n" : "-WRONGPASS invalid credentials\r\n",
                    0);
                MegaCacheClient.Options options = new MegaCacheClient.Options();
                options.host = "127.0.0.1";
                options.port = server.getLocalPort();
                options.password = "secret";
                options.timeout = Duration.ofSeconds(1);
                try (MegaCacheClient client = new MegaCacheClient(options)) {
                    boolean rejected = false;
                    try { client.ping(); }
                    catch (MegaCacheClient.ProtocolException expected) {
                        rejected = protocol;
                    }
                    catch (MegaCacheClient.CommandException expected) {
                        rejected = !protocol;
                    }
                    check(rejected, "AUTH exception type was not preserved");
                }
                responder.join();
            }
        }
    }
    private static void commandDecoderRegressions() throws Exception {
        malformedResponseRegression(
            "PING type", ":1\r\n", MegaCacheClient::ping);
        malformedResponseRegression(
            "GET type", "+value\r\n", client -> client.get("key"));
        malformedResponseRegression(
            "SET shape", bulk("OK"),
            client -> client.set("key", "value".getBytes(StandardCharsets.UTF_8), null));
        malformedResponseRegression(
            "MGET arity", "*1\r\n$5\r\nvalue\r\n",
            client -> client.mget(java.util.List.of("first", "second")));
        malformedResponseRegression(
            "MGET element", "*1\r\n:1\r\n",
            client -> client.mget(java.util.List.of("key")));
        malformedResponseRegression(
            "MSET shape", bulk("OK"),
            client -> client.mset(Map.of(
                "key", "value".getBytes(StandardCharsets.UTF_8))));
        malformedResponseRegression(
            "DEL range", ":-1\r\n", client -> client.delete("key"));
        malformedResponseRegression(
            "EXISTS range", ":2\r\n", client -> client.exists("key"));
        malformedResponseRegression(
            "EXPIRE range", ":2\r\n",
            client -> client.expire("key", Duration.ofSeconds(30)));
        malformedResponseRegression(
            "TTL range", ":-3\r\n", client -> client.ttl("key"));
        malformedResponseRegression(
            "MC.SET shape", ":1\r\n",
            client -> client.put(
                "key", "value".getBytes(StandardCharsets.UTF_8),
                new MegaCacheClient.CachePolicy(), null));
        malformedResponseRegression(
            "MC.LEASE arity", "*1\r\n$5\r\nfresh\r\n",
            client -> client.lease("key"));
        malformedResponseRegression(
            "MC.LEASE element",
            "*4\r\n$5\r\nfresh\r\n$5\r\nvalue\r\n$3\r\nbad\r\n:1\r\n",
            client -> client.lease("key"));
        malformedResponseRegression(
            "MC.LEASE retry", "*2\r\n$7\r\nloading\r\n:-1\r\n",
            client -> client.lease("key"));
        malformedResponseRegression(
            "MC.FETCH JSON", bulk("{"),
            client -> client.fetch("key", "origin", "/path", false, null));
        malformedResponseRegression(
            "MC.FETCH semantics",
            bulk("{\"state\":1,\"origin\":\"origin\"}"),
            client -> client.fetch("key", "origin", "/path", false, null));
        malformedResponseRegression(
            "MC.INVALIDATE range", ":-1\r\n",
            client -> client.invalidate("tag"));
        malformedResponseRegression(
            "MC.STATUS semantics", bulk("{\"degraded\":\"false\"}"),
            MegaCacheClient::status);
        malformedResponseRegression(
            "MC.INVALIDATIONS semantics",
            bulk("{\"cursor\":\"x\",\"epoch\":\"e\",\"generation\":\"bad\"}"),
            client -> {
                Field field = MegaCacheClient.class.getDeclaredField(
                    "lastPollNanos");
                field.setAccessible(true);
                field.setLong(
                    client,
                    System.nanoTime() - Duration.ofDays(3650).toNanos());
                client.getOrLoad(
                    "key", () -> "value".getBytes(StandardCharsets.UTF_8),
                    new MegaCacheClient.CachePolicy());
            });
        malformedResponseRegression(
            "MC.TRACEPARENT shape", bulk("OK"),
            client -> client.setTraceparent("trace"));
    }
    private static Thread respondOnce(
        ServerSocket server, String response, long delayMillis
    ) {
        Thread responder = new Thread(() -> {
            try (Socket socket = server.accept()) {
                socket.getInputStream().read();
                Thread.sleep(delayMillis);
                OutputStream output = socket.getOutputStream();
                output.write(response.getBytes(StandardCharsets.US_ASCII));
                output.flush();
            } catch (Exception error) {
                throw new RuntimeException(error);
            }
        });
        responder.start();
        return responder;
    }
    private static MegaCacheClient deadlineClient(int port) {
        MegaCacheClient.Options options = new MegaCacheClient.Options();
        options.host = "127.0.0.1";
        options.port = port;
        options.timeout = Duration.ofSeconds(1);
        options.invalidationPoll = Duration.ofDays(3650);
        return new MegaCacheClient(options);
    }
    private static MegaCacheClient.CachedValue localValue(
        MegaCacheClient client, String key
    ) throws Exception {
        Field localField = MegaCacheClient.class.getDeclaredField("local");
        localField.setAccessible(true);
        Object local = localField.get(client);
        Method get = local.getClass().getDeclaredMethod("get", String.class);
        get.setAccessible(true);
        return (MegaCacheClient.CachedValue) get.invoke(local, key);
    }
    private static void staleDeadlineRegressions() throws Exception {
        try (ServerSocket server = new ServerSocket(0)) {
            Thread responder = respondOnce(
                server,
                "*3\r\n$5\r\nstale\r\n$5\r\nvalue\r\n:100\r\n",
                60);
            try (MegaCacheClient client = deadlineClient(server.getLocalPort())) {
                MegaCacheClient.CachePolicy policy = new MegaCacheClient.CachePolicy();
                policy.ttl = Duration.ofSeconds(1);
                policy.stale = Duration.ofSeconds(1);
                MegaCacheClient.CachedValue value = client.getOrLoad(
                    "key", () -> "unused".getBytes(StandardCharsets.UTF_8), policy);
                check("stale".equals(value.state), "stale lease response");
                Thread.sleep(55);
                check(localValue(client, "key") == null, "L1 hard deadline exceeded");
            }
            responder.join();
        }
        try (ServerSocket server = new ServerSocket(0)) {
            Thread responder = respondOnce(
                server,
                "*4\r\n$11\r\nstale_lease\r\n$5\r\nstale\r\n"
                    + "$5\r\ntoken\r\n:20\r\n",
                0);
            try (MegaCacheClient client = deadlineClient(server.getLocalPort())) {
                MegaCacheClient.CachePolicy policy = new MegaCacheClient.CachePolicy();
                policy.ttl = Duration.ofSeconds(1);
                policy.stale = Duration.ofSeconds(1);
                boolean failed = false;
                try {
                    client.getOrLoad("key", () -> {
                        try { Thread.sleep(40); }
                        catch (InterruptedException error) {
                            Thread.currentThread().interrupt();
                        }
                        throw new IllegalStateException("load failed");
                    }, policy);
                } catch (IllegalStateException expected) {
                    failed = true;
                }
                check(failed, "expired stale-if-error value was returned");
            }
            responder.join();
        }
    }
    private static void coalescedErrorRegression() throws Exception {
        try (ServerSocket server = new ServerSocket(0)) {
            Thread responder = respondOnce(
                server,
                "*2\r\n$5\r\nlease\r\n$5\r\ntoken\r\n",
                0);
            try (MegaCacheClient client = deadlineClient(server.getLocalPort())) {
                MegaCacheClient.CachePolicy policy =
                    new MegaCacheClient.CachePolicy();
                CountDownLatch entered = new CountDownLatch(1);
                CountDownLatch release = new CountDownLatch(1);
                AssertionError failure = new AssertionError("loader failed");
                AtomicReference<Throwable> leaderFailure =
                    new AtomicReference<>();
                AtomicReference<Throwable> followerFailure =
                    new AtomicReference<>();
                Thread leader = new Thread(() -> {
                    try {
                        client.getOrLoad("key", () -> {
                            entered.countDown();
                            try { release.await(); }
                            catch (InterruptedException error) {
                                Thread.currentThread().interrupt();
                            }
                            throw failure;
                        }, policy);
                    } catch (Throwable error) {
                        leaderFailure.set(error);
                    }
                });
                leader.start();
                check(entered.await(1, java.util.concurrent.TimeUnit.SECONDS),
                    "leader loader did not start");
                Thread follower = new Thread(() -> {
                    try {
                        client.getOrLoad(
                            "key",
                            () -> {
                                throw new AssertionError(
                                    "follower unexpectedly became leader");
                            },
                            policy);
                    } catch (Throwable error) {
                        followerFailure.set(error);
                    }
                });
                follower.start();
                Thread.sleep(20);
                release.countDown();
                leader.join(1000);
                follower.join(1000);
                check(!leader.isAlive(), "leader hung after loader Error");
                check(!follower.isAlive(), "follower hung after loader Error");
                check(leaderFailure.get() == failure,
                    "leader Error semantics changed");
                check(followerFailure.get() == failure,
                    "follower did not receive the leader Error");
            }
            responder.join();
        }
    }
    public static void main(String[] args) throws Exception {
        Map<String, String> v = fixtures();
        try (MegaCacheClient client = client()) {
            check("PONG".equals(client.ping()), "ping");
            client.set(v.get("key"), v.get("value").getBytes(StandardCharsets.UTF_8), null);
            check(v.get("value").equals(new String(client.get(v.get("key")), StandardCharsets.UTF_8)), "get");
            client.mset(Map.of(v.get("second_key"), v.get("second_value").getBytes(StandardCharsets.UTF_8)));
            check(client.mget(java.util.List.of(v.get("key"), v.get("second_key"))).size() == 2, "mget");
            check(client.exists(v.get("key"), v.get("second_key")) == 2, "exists");
            check(client.expire(v.get("key"), Duration.ofSeconds(30)), "expire");
            check(client.ttl(v.get("key")) >= 0, "ttl");
            check(client.delete(v.get("second_key")) == 1, "delete");
            MegaCacheClient.CachePolicy policy = new MegaCacheClient.CachePolicy();
            policy.ttl = Duration.ofSeconds(30); policy.stale = Duration.ofSeconds(30);
            policy.tags = java.util.List.of(v.get("tag"));
            client.put(v.get("key"), v.get("value").getBytes(StandardCharsets.UTF_8), policy, null);
            MegaCacheClient.LeaseResult lease = client.lease(v.get("key"));
            check("fresh".equals(lease.state), "lease");
            check(!lease.expiresIn.isZero() && !lease.staleFor.isZero(), "lease windows");
            check(client.invalidate(v.get("tag")) == 1, "invalidate");
            client.setTraceparent(v.get("traceparent"));
            check(client.fetch(v.get("fetch_key"), v.get("origin"), v.get("path"), false, null).contains(v.get("path")), "fetch");
            check(client.status().contains("\"degraded\":false"), "status");
        }
        try (MegaCacheClient client = client()) {
            client.delete(v.get("l1_key"));
            MegaCacheClient.CachePolicy policy = new MegaCacheClient.CachePolicy();
            policy.ttl = Duration.ofSeconds(30); policy.stale = Duration.ofSeconds(30);
            AtomicInteger calls = new AtomicInteger();
            CountDownLatch start = new CountDownLatch(1);
            Runnable load = () -> {
                try {
                    start.await();
                    MegaCacheClient.CachedValue value = client.getOrLoad(v.get("l1_key"), () -> {
                        calls.incrementAndGet();
                        try { Thread.sleep(20); } catch (InterruptedException error) { Thread.currentThread().interrupt(); }
                        return "loaded".getBytes(StandardCharsets.UTF_8);
                    }, policy);
                    check("loaded".equals(new String(value.value, StandardCharsets.UTF_8)), "L1 value");
                } catch (Exception error) { throw new RuntimeException(error); }
            };
            Thread first = new Thread(load); Thread second = new Thread(load);
            first.start(); second.start(); start.countDown(); first.join(); second.join();
            check(calls.get() == 1, "coalescing");
            try (MegaCacheClient writer = client()) {
                writer.set(v.get("l1_key"), "external".getBytes(StandardCharsets.UTF_8), null);
            }
            MegaCacheClient.CachedValue refreshed = client.getOrLoad(
                v.get("l1_key"),
                () -> { throw new AssertionError("fresh L2 value should avoid loader"); },
                policy
            );
            check("external".equals(new String(refreshed.value, StandardCharsets.UTF_8)), "invalidation");
        }
        protocolLimitRegression("$16777217\r\n");
        protocolLimitRegression("*100001\r\n");
        protocolLimitRegression(
            String.join("", java.util.Collections.nCopies(130, "*1\r\n")) +
            "$1\r\nx\r\n"
        );
        authExceptionRegressions();
        commandDecoderRegressions();
        staleDeadlineRegressions();
        coalescedErrorRegression();
        System.out.println("Java SDK conformance passed");
    }
}

File: sdk/node/test/conformance.test.js
"use strict";

const assert = require("node:assert/strict");
const fs = require("node:fs");
const net = require("node:net");
const test = require("node:test");
const {
  MegaCacheClient, ProtocolError
} = require("../src");

function fixtures() {
  return Object.fromEntries(
    fs.readFileSync(process.env.MEGACACHE_FIXTURES, "utf8")
      .split("\n")
      .filter((line) => line && !line.startsWith("#"))
      .map((line) => line.split("\t"))
  );
}

function decodeCommands(socket, onCommand) {
  let buffered = Buffer.alloc(0);
  socket.on("data", (chunk) => {
    buffered = Buffer.concat([buffered, chunk]);
    for (;;) {
      const lineEnd = buffered.indexOf("\r\n");
      if (lineEnd < 0 || buffered[0] !== 42) return;
      const count = Number(buffered.subarray(1, lineEnd).toString());
      let offset = lineEnd + 2;
      const parts = [];
      for (let index = 0; index < count; index += 1) {
        const lengthEnd = buffered.indexOf("\r\n", offset);
        if (lengthEnd < 0 || buffered[offset] !== 36) return;
        const length = Number(buffered.subarray(offset + 1, lengthEnd).toString());
        const valueStart = lengthEnd + 2;
        const valueEnd = valueStart + length;
        if (buffered.length < valueEnd + 2) return;
        parts.push(buffered.subarray(valueStart, valueEnd).toString());
        offset = valueEnd + 2;
      }
      buffered = buffered.subarray(offset);
      onCommand(parts);
    }
  });
}

test("concurrent first commands wait for connection and AUTH", async () => {
  const seen = [];
  let connections = 0;
  const server = net.createServer((socket) => {
    connections += 1;
    decodeCommands(socket, (parts) => {
      seen.push(parts);
      if (parts[0] === "AUTH") {
        setTimeout(() => socket.write("+OK\r\n"), 40);
      } else {
        socket.write("+PONG\r\n");
      }
    });
  });
  await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
  const client = new MegaCacheClient({
    host: "127.0.0.1",
    port: server.address().port,
    password: "secret",
    timeoutMs: 1000
  });
  try {
    const first = client.ping();
    const second = client.ping();
    await new Promise((resolve) => setTimeout(resolve, 20));
    assert.deepEqual(seen.map((parts) => parts[0]), ["AUTH"]);
    assert.deepEqual(await Promise.all([first, second]), ["PONG", "PONG"]);
    assert.equal(connections, 1);
    assert.deepEqual(
      seen.map((parts) => parts[0]), ["AUTH", "PING", "PING"]
    );
  } finally {
    client.close();
    await new Promise((resolve) => server.close(resolve));
  }
});

test("all command decoders reject malformed responses", async () => {
  const cases = [
    ["ping type", 1, (client) => client.ping()],
    ["get type", "value", (client) => client.get("key")],
    ["set shape", Buffer.from("OK"), (client) => client.set("key", "value")],
    ["mget arity", [Buffer.from("value")],
      (client) => client.mget(["first", "second"])],
    ["mget element", [1], (client) => client.mget(["key"])],
    ["mset shape", Buffer.from("OK"),
      (client) => client.mset([["key", "value"]])],
    ["delete range", -1, (client) => client.delete("key")],
    ["exists range", 2, (client) => client.exists("key")],
    ["expire range", 2, (client) => client.expire("key", 30)],
    ["ttl range", -3, (client) => client.ttl("key")],
    ["put shape", 1, (client) => client.put("key", "value")],
    ["lease arity", [Buffer.from("fresh")], (client) => client.lease("key")],
    ["lease element", [
      Buffer.from("fresh"), Buffer.from("value"), Buffer.from("bad"), 1
    ], (client) => client.lease("key")],
    ["lease retry", [Buffer.from("loading"), -1],
      (client) => client.lease("key")],
    ["fetch JSON", Buffer.from("{"),
      (client) => client.fetch("key", "origin", "/path")],
    ["fetch semantics", Buffer.from('{"state":1,"origin":"origin"}'),
      (client) => client.fetch("key", "origin", "/path")],
    ["invalidate range", -1, (client) => client.invalidate("tag")],
    ["status shape", Buffer.from('{"degraded":"false"}'),
      (client) => client.status()],
    ["invalidations semantics", Buffer.from(
      '{"cursor":"x","epoch":"e","generation":"bad"}'
    ), (client) => client.invalidations()],
    ["traceparent shape", Buffer.from("OK"),
      (client) => client.setTraceparent("trace")]
  ];
  for (const [name, response, operation] of cases) {
    const client = new MegaCacheClient();
    client.command = async () => response;
    await assert.rejects(operation(client), ProtocolError, name);
  }
});

test("RESP2 and MegaCache conformance", {
  skip: !process.env.MEGACACHE_FIXTURES
}, async () => {
  const v = fixtures();
  const client = new MegaCacheClient({
    host: process.env.MEGACACHE_CONFORMANCE_HOST,
    port: Number(process.env.MEGACACHE_CONFORMANCE_PORT),
    invalidationPollMs: 0
  });
  try {
    assert.equal(await client.ping(), "PONG");
    assert.equal(await client.set(v.key, v.value), "OK");
    assert.equal((await client.get(v.key)).toString(), v.value);
    assert.equal(await client.mset([[v.second_key, v.second_value]]), "OK");
    assert.deepEqual((await client.mget([v.key, v.second_key])).map(String), [v.value, v.second_value]);
    assert.equal(await client.exists(v.key, v.second_key), 2);
    assert.equal(await client.expire(v.key, 30), 1);
    assert.ok(await client.ttl(v.key) >= 0);
    assert.equal(await client.delete(v.second_key), 1);
    await client.put(v.key, v.value, { ttlSeconds: 30, staleSeconds: 30, tags: [v.tag] });
    const lease = await client.lease(v.key);
    assert.equal(lease.state, "fresh");
    assert.ok(lease.expiresInMs > 0);
    assert.ok(lease.staleForMs > 0);
    assert.equal(await client.invalidate(v.tag), 1);
    assert.equal(await client.get(v.key), null);
    await client.setTraceparent(v.traceparent);
    const fetched = await client.fetch(v.fetch_key, v.origin, v.path);
    assert.equal(fetched.value, v.path);
    assert.equal((await client.status()).degraded, false);
  } finally {
    client.close();
  }
});

test("L1 coalesces concurrent loaders", {
  skip: !process.env.MEGACACHE_FIXTURES
}, async () => {
  const v = fixtures();
  const client = new MegaCacheClient({
    host: process.env.MEGACACHE_CONFORMANCE_HOST,
    port: Number(process.env.MEGACACHE_CONFORMANCE_PORT),
    invalidationPollMs: 0
  });

  let calls = 0;
  try {
    await client.delete(v.l1_key);
    const loader = async () => {
      calls += 1;
      await new Promise((resolve) => setTimeout(resolve, 20));
      return "loaded";
    };
    const results = await Promise.all([
      client.getOrLoad(v.l1_key, loader, { ttlSeconds: 30, staleSeconds: 30 }),
      client.getOrLoad(v.l1_key, loader, { ttlSeconds: 30, staleSeconds: 30 })
    ]);
    assert.equal(calls, 1);
    assert.deepEqual(results.map((item) => item.value.toString()), ["loaded", "loaded"]);
    const writer = new MegaCacheClient({
      host: process.env.MEGACACHE_CONFORMANCE_HOST,
      port: Number(process.env.MEGACACHE_CONFORMANCE_PORT)
    });
    await writer.set(v.l1_key, "external");
    writer.close();
    const refreshed = await client.getOrLoad(v.l1_key, async () => {
      throw new Error("fresh L2 value should avoid the loader");
    }, { ttlSeconds: 30, staleSeconds: 30 });
    assert.equal(refreshed.value.toString(), "external");
  } finally {
    client.close();
  }
});

test("zero lease freshness is not admitted and absent windows use policy", async () => {
  const client = new MegaCacheClient({ invalidationPollMs: 60000 });
  client.lastPoll = Date.now();
  client.lease = async () => ({
    state: "fresh", value: Buffer.from("zero"), expiresInMs: 0, staleForMs: 0
  });
  assert.equal((await client.getOrLoad("zero", async () => "unused")).value.toString(), "zero");
  assert.equal(client.local.get("zero"), null);

  client.lease = async () => ({ state: "fresh", value: Buffer.from("legacy") });
  await client.getOrLoad("legacy", async () => "unused", { ttlSeconds: 30, staleSeconds: 30 });
  assert.equal(client.local.get("legacy").state, "fresh");
});

test("in-flight reads cannot admit across a successful mutation", async () => {
  const client = new MegaCacheClient({ invalidationPollMs: 60000 });
  client.lastPoll = Date.now();
  let release;
  client.lease = () => new Promise((resolve) => { release = resolve; });
  const loading = client.getOrLoad("key", async () => "unused", {
    ttlSeconds: 30, staleSeconds: 30
  });
  await new Promise((resolve) => setImmediate(resolve));
  client.mutationSucceeded();
  release({ state: "fresh", value: Buffer.from("old"), expiresInMs: 30000, staleForMs: 30000 });
  assert.equal((await loading).value.toString(), "old");
  assert.equal(client.local.get("key"), null);
});

test("lease elapsed time is removed from the L1 hard deadline", async () => {
  const client = new MegaCacheClient({ invalidationPollMs: 60000 });
  client.lastPoll = Date.now();
  client.lease = async () => {
    await new Promise((resolve) => setTimeout(resolve, 60));
    return {
      state: "stale", value: Buffer.from("value"), staleForMs: 100
    };
  };
  await client.getOrLoad("key", async () => "unused", {
    ttlSeconds: 1, staleSeconds: 1
  });
  await new Promise((resolve) => setTimeout(resolve, 55));
  assert.equal(client.local.get("key"), null);
});

test("stale-if-error does not cross the lease hard deadline", async () => {
  const client = new MegaCacheClient({ invalidationPollMs: 60000 });
  client.lastPoll = Date.now();
  client.lease = async () => ({
    state: "stale_lease",
    value: Buffer.from("stale"),
    leaseToken: "token",
    staleForMs: 20
  });
  await assert.rejects(
    client.getOrLoad("key", async () => {
      await new Promise((resolve) => setTimeout(resolve, 40));
      throw new Error("load failed");
    }, { ttlSeconds: 1, staleSeconds: 1 }),
    /load failed/
  );
});

test("failed mutations retain L1 entries and generation", async () => {
  const client = new MegaCacheClient();
  client.local.put("key", "value", 30, 30);
  const generation = client.generation;
  client.command = async () => { throw new Error("rejected"); };
  await assert.rejects(client.set("other", "value"), /rejected/);
  assert.equal(client.generation, generation);
  assert.equal(client.local.get("key").value.toString(), "value");
});

test("request traceparents are isolated across asynchronous scopes", async () => {
  const client = new MegaCacheClient();
  const seen = [];
  client.command = async (...parts) => {
    seen.push(parts[parts.indexOf("TRACEPARENT") + 1]);
    return Buffer.from('{"state":"fresh","origin":"catalog"}');
  };
  await Promise.all([
    client.withTraceparent("trace-a", async () => {
      await new Promise((resolve) => setTimeout(resolve, 10));
      await client.fetch("a", "catalog", "/a");
    }),
    client.withTraceparent("trace-b", async () => client.fetch("b", "catalog", "/b"))
  ]);
  assert.deepEqual(new Set(seen), new Set(["trace-a", "trace-b"]));
});

test("protocol limit failures poison and reset the connection", async () => {
  const responses = [
    "$16777217\r\n",
    "*100001\r\n",
    "*1\r\n".repeat(130) + "$1\r\nx\r\n"
  ];
  for (const response of responses) {
    let connections = 0;
    const server = net.createServer((socket) => {
      connections += 1;
      socket.once("data", () => {
        if (connections === 1) socket.write(response);
        else socket.write("+PONG\r\n");
      });
    });
    await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
    const client = new MegaCacheClient({
      host: "127.0.0.1", port: server.address().port, timeoutMs: 1000
    });
    try {
      await assert.rejects(client.ping(), ProtocolError);
      assert.equal(client.connection.buffer.length, 0);
      assert.equal(await client.ping(), "PONG");
    } finally {
      client.close();
      await new Promise((resolve) => server.close(resolve));
    }
  }
});
